node/narwhal/src/helpers/codec.rs

```diff
@@ -84,3 +84,311 @@ impl<N: Network> Decoder for EventCodec<N> {
+                let chunked_plaintext_msg: Vec<_> = bytes.chunks(MAX_MESSAGE_LEN - 16).collect();
+                let num_chunks = chunked_plaintext_msg.len() as u64;
+
+                // Encrypt the resulting bytes with Noise.
+                let encrypted_chunks: Vec<io::Result<Vec<u8>>> = chunked_plaintext_msg
+                    .into_par_iter()
+                    .enumerate()
+                    .map(|(nonce_offset, plaintext_chunk)| {
+                        let mut buffer = vec![0u8; MAX_MESSAGE_LEN];
+                        let len = noise
+                            .state
+                            .write_message(noise.tx_nonce + nonce_offset as u64, plaintext_chunk, &mut buffer)
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        buffer.truncate(len);
+
+                        Ok(buffer)
+                    })
+                    .collect();
+
+                let mut buffer = BytesMut::new();
+                // Set the message type flag.
+                buffer.put_u8(message_type as u8);
+
+                for chunk in encrypted_chunks {
+                    buffer.extend_from_slice(&chunk?)
+                }
+
+                noise.tx_nonce += num_chunks;
+
+                buffer
+            }
+        };
+
+        // Encode the resulting ciphertext using the length-delimited codec.
+        self.codec.encode(ciphertext.freeze(), dst)
+    }
+}
+
+impl<N: Network> Decoder for NoiseCodec<N> {
+    type Error = io::Error;
+    type Item = EventOrBytes<N>;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // Decode the ciphertext with the length-delimited codec.
+        let (flag, bytes) = if let Some(mut bytes) = self.codec.decode(src)? {
+            let flag =
+                MessageType::try_from(bytes.get_u8()).map_err(|e| Self::Error::new(io::ErrorKind::InvalidData, e))?;
+            (flag, bytes)
+        } else {
+            return Ok(None);
+        };
+
+        let msg = match self.noise_state {
+            NoiseState::Handshake(ref mut noise) => {
+                // Ignore any messages before the noise handshake has completed.
+                if let MessageType::Event = flag {
+                    return Ok(None);
+                }
+
+                // Decrypt the ciphertext in handshake mode.
+                let mut buffer = [0u8; MAX_MESSAGE_LEN];
+                let len = noise.read_message(&bytes, &mut buffer).map_err(|_| io::ErrorKind::InvalidData)?;
+
+                Some(EventOrBytes::Bytes(Bytes::copy_from_slice(&buffer[..len])))
+            }
+
+            NoiseState::PostHandshake(ref mut noise) => {
+                // Ignore raw bytes after the noise handshake has completed.
+                if let MessageType::Bytes = flag {
+                    return Ok(None);
+                }
```

Comment:
hmm, now that I think about it, do we even need this flag if we always send bytes during the handshake and events post-handshake?

Reply:
we would still need the `enum EventOrBytes`, but not necessarily the leading byte or its `enum`